### DISCUSSION TITLE
pkcs12: -passin file:xxx and -passout file:xxx cannot read from the same file

### DISCUSSION THREAD

**Discussion by targetdrone** [openssl/openssl#21763] [Q&A] [answered]
## Issue

I'm trying to use the openssl command to create a PKCS#12 keystore from a private key, certificate, and CA certificate. All three files are PEM formatted.

I'm using `openssl pkcs12 -export` by passing in the same password file for both decrypting the private.key file and encrypting the private key in the new PKCS#12 keystore,  It fails and I get the following error messages:

```
Error reading password from BIO
Error getting passwords
```

### Version

I reproduced this using `OpenSSL 3.1.1 30 May 2023 (Library: OpenSSL 3.1.1 30 May 2023)` on Windows 10. The pre-built image was obtained from Chocolatey here: https://community.chocolatey.org/packages/openssl/3.1.1

I have received reports from other people using Linux and MacOS X that they are encountering the same issue, but I did not try to reproduce it on other platforms or with other versions of OpenSSL.

## Steps to Recreate

I first create a strong password, then a private key and CSR, get the CSR signed, and then attempt to build a PKCS#12 keystore.

### Generate a strong random password
```
openssl rand -base64 15 > pw.txt
```

### Generate a private key and CSR, encrypting the private key with the generated password
```
openssl req -new -newkey rsa:2048 -keyform PEM -keyout private.key  -passout file:pw.txt -subj "/CN=host.example.com /O=organization/L=locality/ST=state/C=US" -addext "subjectAltName=DNS:host.example.com" -outform PEM -out request.csr
```
This worked.  I sent the CSR for signing and stored the certificate in a file named certificate.pem. It was signed by the CA whose cert is in ca.pem.

### Create a PKCS#12 keystore:
```
openssl pkcs12 -export -certfile ca.pem -in certificate.pem -inkey private.key -passin file:pw.txt -passout file:pw.txt -out keystore.pfx
```

This failed.  The result of this call is this error:
```
Error reading password from BIO
Error getting passwords
```

## Workaround
If I make a copy of the pw.txt file, I can successfully make the call by passing two different files for `-passin` and `-passout`:
```
copy pw.txt pw2.txt
openssl pkcs12 -export -certfile ca.pem -in certificate.pem -inkey private.key -passin file:pw.txt -passout file:pw2.txt -out keystore.pfx
```

**Comment by t8m** [comment, marked as answer]
The input and output passwords in such case have to be on separate lines. I.e., duplicated in your case.

This is documented in openssl-passphrase-options(1) manual page. https://www.openssl.org/docs/man3.0/man1/openssl-passphrase-options.html
